### PR TITLE
Add fare attributes and fare rules

### PR DESCRIPTION
**Pull request by LeoFrachet**

### CODE REVIEW

File: gtfs.js [hunk 1]
```diff
@@ -75,9 +100,15 @@ function forEachItem(gtfs, tableName, iterator) {
   }
 
   const indexedTable = gtfs.getIndexedTable(tableName);
+  const indexKeys = gtfs._schema.indexKeysByTableName[tableName];
   const deepness = gtfs._schema.deepnessByTableName[tableName];
 
-  if (deepness === 1) {
+  if (indexKeys.singleton) {
```
Comment (gcamp) [inline gtfs.js]: I don't think singleton is the right name. `isIndexless` maybe?

File: gtfs.js [hunk 2]
```diff
@@ -1155,15 +1219,155 @@ class Gtfs {
...
+   *
+   * @returns {number}
+   */
+  getNumberOfFareAttributes() { return getNumberOfItemsInTable(this, 'fare_attributes'); }
+
+  /**
+   * Get the fareAttribute using its fare_id.
+   *
+   * WARNING: Will return the fareAttribute which is indexed with the fare_id passed as argument. If the internal value
+   * of the fareAttribute's fare_id has been changed but not it's indexing, the result will be wrong.
```
Comment (gcamp) [inline gtfs.js]: FYI got bitten by this in an other context on trips (trip_id changed but index not changed). Maybe we should consider having some regular sanity check, at least at export time.

File: tests/test_fare_attributes.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+describe('Tests on GTFS fare attributes', () => {
+  it('Tests on gtfs.getNumberOfFareAttributes(), .add…(), .remove…() and .getFareAttributeWithId', (done) => {
+    const path = `${__dirname}/samples/1`;
+    const gtfs = new Gtfs(path);
+    const fareAttribute1 = buildFareAttribute(1);
+    const fareAttribute2 = buildFareAttribute(2);
+    const fareAttribute3 = buildFareAttribute(3);
+
+    expect(gtfs.getNumberOfFareAttributes()).to.equal(0);
+    expect(gtfs.getFareAttributeWithId('fareAttribute.fare_id.1') !== undefined).to.equal(false);
```
Comment (lvancraen) [inline tests/test_fare_attributes.js]: Maybe it's just me, but I find it easier to read ```expect(gtfs.getFareAttributeWithId('fareAttribute.fare_id.1') === undefined).to.equal(true);```

Comment (lvancraen) [reply tests/test_fare_attributes.js]: Though I can see why you did that with the other tests that follow

File: tests/test_fare_attributes.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
+'use strict';
+
+const { expect } = require('chai');
+const { Gtfs } = require('../index');
+
+describe('Tests on GTFS fare attributes', () => {
+  it('Tests on gtfs.getNumberOfFareAttributes(), .add…(), .remove…() and .getFareAttributeWithId', (done) => {
```
Comment (lvancraen) [inline tests/test_fare_attributes.js]: What happens if you add/remove a fare that doesn't exist?

Comment (LeoFrachet) [reply tests/test_fare_attributes.js]: Removing a fare that doesn't exist won't do anything. It is like deleting an undefined key.

Adding a fare which doesn't exist should not work.

**Pull request merged**